man/dnstable_dump.1
'\" t
.\"     Title: dnstable_dump
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 08/06/2021
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "DNSTABLE_DUMP" "1" "08/06/2021" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
dnstable_dump \- dump dnstable data file to text or JSON
.SH "SYNOPSIS"
.sp
\fBdnstable_dump\fR [\fB\-\-raw\fR] [\fB\-\-json\fR] \fB\-\-rrset\fR \fIfilename\fR
.sp
\fBdnstable_dump\fR [\fB\-\-raw\fR] [\fB\-\-json\fR] \fB\-\-rdata\fR \fIfilename\fR
.SH "DESCRIPTION"
.sp
Dumps a dnstable data file to stdout, either in an ad\-hoc text format (the default), or in JSON format\&. Specifying one of \fB\-\-rrset\fR or \fB\-\-rdata\fR is required in order to select which section of the data file to dump\&.
.SH "OPTIONS"
.PP
\fB\-R\fR, \fB\-\-raw\fR
.RS 4
When using JSON format for \-\-rdata_full output, add a hex representation of the raw rdata\&.
.RE
.PP
\fB\-j\fR, \fB\-\-json\fR
.RS 4
Use JSON format for output\&. One JSON\-formatted entry per line will be printed\&.
.RE
.PP
\fB\-r\fR, \fB\-\-rrset_full\fR
.RS 4
Dump the full
\fBRRset\fR
section of the data file from
\fBENTRY_TYPE_RRSET\fR
records\&.
.RE
.PP
\fB\-d\fR, \fB\-\-rdata_full\fR
.RS 4
Dump the full
\fBRdata\fR
section of the data file from
\fBENTRY_TYPE_RDATA\fR
records\&.
.RE
.PP
\fB\-\-rrset_names\fR
.RS 4
Dump the rrset names from
\fBENTRY_TYPE_RRSET_NAME_FWD\fR
records\&.
.RE
.PP
\fB\-\-rdata_names\fR
.RS 4
Dump the rdata names from
\fBENTRY_TYPE_RDATA_NAME_REV\fR
records\&.
.RE
.PP
\fB\-t\fR, \fB\-\-time_range\fR
.RS 4
Output time range metadata from
\fBENTRY_TYPE_TIME_RANGE\fR
records\&.
.RE
.PP
\fB\-s\fR, \fB\-\-source_info_entries\fR
.RS 4
Output source metadata from
\fBENTRY_TYPE_SOURCE_INFO\fR
records\&.
.RE
.PP
\fB\-v\fR, \fB\-\-version_entries\fR
.RS 4
Output version metadata from
\fBENTRY_TYPE_VERSION\fR
records\&.
.RE
.SH "EXAMPLES"
.sp
$ dnstable_dump \-R \-j \-d /path/to/dns\&.mtbl
.sp
$ dnstable_dump \-r /path/to/dns\&.mtbl
.sp
$ dnstable_dump \-\-rdata_names /path/to/dns\&.mtbl


ChangeLog
dnstable (0.14.0)

  * Add "+" wildcard character to match single label at beginning or end of name.
  * Add support for case-sensitive rrset, rdata name queries.
  * Multiple performance improvements for time filtered queries.
  * Improve performance of rrtype filtered queries through early filtering.
  * Add query statistics API.
  * Add support for source info metadata.
  * dnstable_lookup: Add "case sensitive" (-C) query flag.
  * dnstable_lookup: Add "count results" (-n) flag.
  * dnstable_lookup: Add "time limit" (-t) option.
  * dnstable_lookup: Add stats output (-s) and stats interval (-i) options.
  * Fix building shared libraries on windows/Cygwin.

dnstable (0.13.2)

  * Fix bug handling short rdata entry keys in IPv6 range queries.
  * Efficiency improvements in string handling code and IP range queries.

dnstable (0.13.1)

  * Add RP, NXT, and NSEC to rdata name type list for matching rdata names.

dnstable (0.13.0)

  * Update dnstable to 0.13.0 version
  * Updated dependency to libwdns 0.11.0
  * Updated dependency to libmtbl 1.5.0
  * dnstable_lookup: add new command-line options -A, -a, -B, -b, and -c for time-fenced lookups.
  * Optimize time filtered searches on fileset sources with a filtering query using the subset of files which may contribute to matching results.
  * Fix duplication of results in unaggregated name prefix or rdata name suffix searches.
  * Fix improper skipping of results in IP range or prefix searches.
  * Fix return of non-name results in rdata name searches.
  * Fix decode_rdata() rdata length for Big Endian.

dnstable (0.12.0)

  * Implement indexing using RRtypes for ENTRY_TYPE_RRSET_NAME_FWD
    and ENTRY_TYPE_RDATA_NAME_REV entries. This uses RFC6840 RRtype
    bitmaps as the primary data structure.

  * Added support for dnstable entries encoding version information and
    time metadata.

  * The version entries, time metadata, and rrtype index information are
    generated by an updated dnstable-convert. Files generated by
    the old dnstable-convert will be processed as they were with the
    previous dnstable versions i.e., the new entry types and rrset index
    information will be absent.

  * Use ./configure --with-coverage option to build with code coverage and
    make targets: clean-coverage to remove the coverage data and results
    and report-coverage to generate LCOV report (after running the code
    such as with "make check").

  * Add -R flag to dnstable_dump and dnstable_lookup to add raw rdata
    representation to outputs.

  * Add long options to dnstable_dump to fix bug in not accepting some
    options properly.

  * Fix dnstable_lookup usage output.

  * Refactor time formatters and fix compiler warnings.

  * Ignore entries which fail decoding for use with legacy empty data.

  * Miscellaneous documentation improvements.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Thu, 14 Oct 2021 12:35:12 -0400

dnstable (0.11.2)

  * Fix memory leaks while skipping rows of data during offset processing.

  * Fix minor leak in dnstable_lookup printing.

  * Add more dnstable_lookup tests.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Wed, 26 Jun 2019 22:00:00 +0000

dnstable (0.11.1)

  * Change dnstable_query_set_skip to ..._offset. Note that the
    ABI library revision was not bumped.

  * Change dnstable_lookup parameter for offset from -s to -O

  * Added unit tests for some exposed dnstable functionality, including
    for filter and query timeouts.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Mon, 17 Jun 2019 13:44:34 +0000

dnstable (0.11.0)

  * Fix bug in raw rdata queries with rrtype specified.

  * Add functionality for unaggregated query results.

  * Add ability to skip query results.

  * Add options for date and rdata presentation in JSON formatted entries.

  * Treat RRTYPE "TYPE255" as an alias for "ANY", since 255 is the code for "ANY".

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Fri, 29 Mar 2019 12:34:07 -0500

dnstable (0.10.2)

  * Fix ordering for wildcard, processing continguous memory first.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Thu, 06 Sep 2018 18:37:23 -0400

dnstable (0.10.1)

  * Check for endian.h headers.

  * Add sys/socket.h include.

  * Expunge protobuf-c references.

  * Correct libyajl include path.

 -- Farsight Security, Inc <software@farsightsecurity.com>  Mon, 18 Jun 2018 15:50:31 -0400

dnstable (0.10.0)

  * Remove dnstable_convert into a separate repo to simplify
    dependencies. https://github.com/farsightsec/dnstable-convert/

  * Add generated man pages to source.

  * Document dnstable_entry_set_iszone().

  * Improve dnstable_reader(3) documentation.

  * Add JSON output (-j option) to dnstable_lookup.

  * Add test cases based on dnstable_lookup.

  * Only encode RDATA so the hostname is at the beginning for NS, CNAME,
    DNAME, PTR, MX, and SRV record types.

  * Fix byteorder macros for macOS.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Tue, 05 Jun 2018 13:35:03 +0000

dnstable (0.9.1)

  * Fix strange values in data value triplets (#16). A 'count' field that has
    the value 0 is silently rewritten to 1, and a 'time_last' field that has
    the value 0 is silently set to be the same as the 'time_first' field.

 -- Robert Edmonds <edmonds@fsi.io>  Tue, 19 Apr 2016 18:39:47 -0400

dnstable (0.9.0)

  [ Henry Stern ]
  * New "query timeout" feature which allows for a configurable timeout on the
    execution of certain types of queries. Introduces new function
    dnstable_query_set_timeout() and new result code dnstable_res_timeout
    (#8).

  * New "time fencing" feature which filters based on 'time_first' and
    'time_last' values. Introduces new enum dnstable_filter_parameter_type and
    new function dnstable_query_set_filter_parameter() (#9).

  * Performance enhancement for some IP range and prefix searches (#11).


src/dnstable_dump.c
/*
 * Copyright (c) 2023-2024 DomainTools LLC
 * Copyright (c) 2012, 2014-2015, 2020, 2021 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <assert.h>
#include <errno.h>
#include <inttypes.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include <dnstable.h>
#include <mtbl.h>

#include "libmy/argv.h"
#include "libmy/print_string.h"

static bool g_json;
static bool g_add_raw;
static bool g_rrset;
static bool g_rdata;
static bool g_rrset_names;
static bool g_rdata_names;
static bool g_time_range;
static bool g_version;
static bool g_source_info;
static char *g_fname;

static struct dnstable_formatter *fmt = NULL;

static argv_t args[] = {
	{ 'j',	"json",
		ARGV_BOOL,
		&g_json,
		NULL,
		"output in JSON format (default: text)" },

	{ 'R',	"raw",
		ARGV_BOOL,
		&g_add_raw,
		NULL,
		"add raw rdata representation to --rdata_full" },

	{ 'r',	"rrset_full",
		ARGV_BOOL,
		&g_rrset,
		NULL,
		"output rrset full records" },

	{ ARGV_ONE_OF },

	{ 'd',	"rdata_full",
		ARGV_BOOL,
		&g_rdata,
		NULL,
		"output rdata full records" },

	{ ARGV_ONE_OF },

	{ '\0', "rrset_names",
		ARGV_BOOL,
		&g_rrset_names,
		NULL,
		"output rrset names" },

	{ ARGV_ONE_OF },

	{ '\0', "rdata_names",
		ARGV_BOOL,
		&g_rdata_names,
		NULL,
		"output rdata names" },

	{ ARGV_ONE_OF },

	{ 't', "time_range",
		ARGV_BOOL,
		&g_time_range,
		NULL,
		"output time range metadata" },

	{ARGV_ONE_OF },

	{ 'v', "version_entries",
		ARGV_BOOL,
		&g_version,
		NULL,
		"output version entries" },

	{ ARGV_ONE_OF },

	{ 's', "source_info_entries",
		ARGV_BOOL,
		&g_source_info,
		NULL,
		"output source info entries" },

	{ ARGV_MAND, NULL,
		ARGV_CHAR_P,
		&g_fname,
		"filename",
		"input file" },

	{ ARGV_LAST }
};

static void
print_entry(struct dnstable_entry *ent)
{
	switch (dnstable_entry_get_type(ent)) {
	case DNSTABLE_ENTRY_TYPE_RRSET:
	case DNSTABLE_ENTRY_TYPE_RRSET_NAME_FWD:
	case DNSTABLE_ENTRY_TYPE_RDATA:
	case DNSTABLE_ENTRY_TYPE_RDATA_NAME_REV:
	case DNSTABLE_ENTRY_TYPE_TIME_RANGE:
	case DNSTABLE_ENTRY_TYPE_SOURCE_INFO:
	case DNSTABLE_ENTRY_TYPE_VERSION: {
		char *s;
		if (g_json)
			s = dnstable_entry_format(fmt, ent);
		else
			s = dnstable_entry_to_text(ent);
		if (s != NULL) {
			fputs(s, stdout);
			if (!(!g_json &&
			      dnstable_entry_get_type(ent) == DNSTABLE_ENTRY_TYPE_RDATA))
			{
				putchar('\n');
			}
			free(s);
		}
	}
	}
}

static void
do_dump(struct dnstable_iter *it)
{
	struct dnstable_entry *ent;

	while (dnstable_iter_next(it, &ent) == dnstable_res_success) {
		assert(ent != NULL);
		print_entry(ent);
		dnstable_entry_destroy(&ent);
	}
}

int
main(int argc, char **argv)
{
	struct mtbl_reader *m_reader;
	struct dnstable_reader *d_reader;
	struct dnstable_iter *d_it = NULL;

	argv_process(args, argc, argv);

	if (g_add_raw) {
		if (!(g_rdata || g_rrset)) {
			fprintf(stderr,
				"dnstable_dump: adding raw rdata only supported with --r or -d output formats\n");
			exit(EXIT_FAILURE);
		} else if (!g_json) {
			fprintf(stderr, "dnstable_dump: adding raw rdata only supported with json output format\n");
			exit(EXIT_FAILURE);
		}
	}

	m_reader = mtbl_reader_init(g_fname, NULL);
	if (m_reader == NULL) {
		perror("mtbl_reader_init");
		fprintf(stderr, "dnstable_dump: unable to open database file %s\n",
			g_fname);
		exit(EXIT_FAILURE);
	}

	d_reader = dnstable_reader_init(mtbl_reader_source(m_reader));

	if (g_rrset) {
		d_it = dnstable_reader_iter_rrset(d_reader);
	} else if (g_rdata) {
		d_it = dnstable_reader_iter_rdata(d_reader);
	} else if (g_rrset_names) {
		d_it = dnstable_reader_iter_rrset_names(d_reader);
	} else if (g_rdata_names) {
		d_it = dnstable_reader_iter_rdata_names(d_reader);
	} else if (g_time_range) {
		d_it = dnstable_reader_iter_time_range(d_reader);
	} else if (g_version) {
		d_it = dnstable_reader_iter_version(d_reader);
	} else if (g_source_info) {
		d_it = dnstable_reader_iter_source_info(d_reader);
	}

	fmt = dnstable_formatter_init();
	dnstable_formatter_set_output_format(fmt, dnstable_output_format_json);
	dnstable_formatter_set_date_format(fmt, dnstable_date_format_unix);
	dnstable_formatter_set_raw_rdata(fmt, g_add_raw);

	assert(d_it != NULL);
	do_dump(d_it);

	dnstable_formatter_destroy(&fmt);

	dnstable_iter_destroy(&d_it);
	dnstable_reader_destroy(&d_reader);
	mtbl_reader_destroy(&m_reader);

	argv_cleanup(args);

	return (EXIT_SUCCESS);
}
